### ISSUE TITLE
Collect timeseries from workers, keep in scheduler and plot

### ISSUE THREAD

**Issue by mrocklin** [dask/distributed#2117]
Currently we record the cpu and memory usage from workers from their heartbeats.  These end up on the main status page of the dashboard as bar charts.

It might be useful to extend this in two ways

1.  Have a separate view that includes these as time series.  Additionally, to support the 1000 worker case we'll need to actually plot something like quantiles instead of all of the values beyond a certain amount
2.  Allow users to specify functions that they would like to register to also be published with these time series.  This would allow people to quickly build things like diagnostics for special hardware like GPUs

cc @lesteve 

###  Register functions

Currently in `Worker.heartbeat` we include all of the values in `**self.monitor.recent()`, which includes things like CPU and memory use.  This gets collected into the WorkerState's `info` attribute on the scheduler.  We can put other things in here as well.

So the worker probably also needs some registry of functions to run along with `self.monitor.recent()`

```python
                response = yield self.scheduler.heartbeat_worker(
                    address=self.contact_address,
                    name=self.name,
                    now=time(),
                    memory_limit=self.memory_limit,
                    executing=len(self.executing),
                    in_memory=len(self.data),
                    ready=len(self.ready),
                    in_flight=len(self.in_flight_tasks),
                    **self.monitor.recent()
                    **{k: func(self) for k, func in self.diagnostic_functions})  # <<--- this is new!
```

Then we will need to have some mechanism for users to add these diagnostic functions.  This can be done in a few ways.  I'll try to write more in a bit.  Short term when testing I would just hardcode a couple of these in a test.

**Comment by mrocklin** [comment]
Here is a simple test to verify that we can register a function on the workers and have that be collected on the scheduler.


```python
@gen_cluster(client=True)
def test_diagnostic_functions(c, s, a, b):
    def myfunc(worker):
        return 123 

    a.diagnostic_functions['foo'] = myfunc
    b.diagnostic_functions['foo'] = myfunc
    
    while 'foo' not in s.workers[a.address].info:
        yield gen.sleep(0.01)

    assert s.workers[a.address].info['foo'] = 123
```

**Comment by mrocklin** [comment]
Closed by #2169
